From bf9e5ceb4ac0afc97d2d44671ff08d460947b7d9 Mon Sep 17 00:00:00 2001
From: David Fort <contact@hardening-consulting.com>
Date: Fri, 16 Nov 2018 12:40:56 +0100
Subject: [PATCH] Add a workaround for disconnections with windows 2k16 server

This disconnection happens with a lot of traffic when using the command line flags
/gdi:hw /bpp:32 -bitmap-cache -offscreen-cache -glyph-cache /network:lan +fonts +aero +window-drag +menu-anims +themes +wallpaper
The connection ends with a Set Error PDU saying that the graphical server part failed.

Sponsored by: Rangee GmbH (http://www.rangee.com)
---
 client/common/client.c | 28 ++++++++++++++++++----------
 1 file changed, 18 insertions(+), 10 deletions(-)

diff --git a/client/common/client.c b/client/common/client.c
index 7b2c54d882..ae673b4392 100644
--- a/client/common/client.c
+++ b/client/common/client.c
@@ -680,17 +680,25 @@ BOOL client_auto_reconnect_ex(freerdp* instance, BOOL(*window_events)(freerdp* i
 	settings = instance->settings;
 	maxRetries = settings->AutoReconnectMaxRetries;
 
-	/* Only auto reconnect on network disconnects. */
-	if (freerdp_error_info(instance) != 0)
-		return FALSE;
-
-	/* A network disconnect was detected */
-	WLog_INFO(TAG, "Network disconnect!");
-
-	if (!settings->AutoReconnectionEnabled)
+	switch (freerdp_error_info(instance))
 	{
-		/* No auto-reconnect - just quit */
-		return FALSE;
+		case 0:
+			/* A network disconnect was detected */
+			WLog_INFO(TAG, "Network disconnect!");
+
+			if (!settings->AutoReconnectionEnabled)
+			{
+				/* No auto-reconnect - just quit */
+				return FALSE;
+			}
+			break;
+		case ERRINFO_GRAPHICS_SUBSYSTEM_FAILED:
+			WLog_INFO(TAG, "Probably hitting bug with 2k16 and 2k12 servers, auto-reconnecting...");
+			if (!maxRetries)
+				maxRetries = 4;
+			break;
+		default:
+			return FALSE;
 	}
 
 	/* Perform an auto-reconnect. */
